tools/lightning/csv.md

```diff
@@ -0,0 +1,222 @@
+
+### `header`
+
+- Whether *all* CSV files contain a header row.
+- If `header` is true, the first row will be used as the
+*column names*. If `header` is false, the first row is not special and treated as an ordinary data row.
+
+### `not-null` and `null`
+
+- The `not-null` setting controls whether all fields are non-nullable.
+- If `not-null` is false, the
+string specified by `null` will be transformed to the SQL NULL instead of a concrete value.
+- Quoting will not affect whether a field is null.
+
+    For example, with the CSV file:
+
+    ```csv
+    A,B,C
+    \N,"\N",
+    ```
+
+    In the default settings (`not-null = false; null = '\N'`), the columns `A` and `B` are both
+converted to NULL after importing to TiDB. The column `C` is simply the empty string `''` but not
+NULL.
+
+### `backslash-escape`
+
+- Whether to interpret backslash escapes inside fields.
+- If `backslash-escape` is true, the following sequences are
+recognized and transformed:
+
+    | Sequence | Converted to             |
+    |----------|--------------------------|
+    | `\0`     | Null character (U+0000)  |
+    | `\b`     | Backspace (U+0008)       |
+    | `\n`     | Line feed (U+000A)       |
+    | `\r`     | Carriage return (U+000D) |
+    | `\t`     | Tab (U+0009)             |
+    | `\Z`     | Windows EOF (U+001A)     |
+
+    In all other cases (e.g. `\"`) the backslash is simply stripped, leaving the next character (`"`)
+in the field.
+
+- Quoting will not affect whether backslash escapes are interpreted.
+
+- Corresponds to the `FIELDS ESCAPED BY '\'` option in the `LOAD DATA` statement.
+
+### `trim-last-separator`
+
+- Treats the field `separator` as a terminator, and removes all trailing separators.
+
+    For example, with the CSV file:
+
+    ```csv
+    A,,B,,
+    ```
+
+- When `trim-last-separator = false`, this is interpreted as a row of 5 fields `('A', '', 'B', '', '')`.
+- When `trim-last-separator = true`, this is interpreted as a row of 3 fields `('A', '', 'B')`.
+
+### Non-configurable options
+
+Lightning does not support every option supported by the `LOAD DATA` statement. Some examples:
+
+* The line terminator must only be CR (`\r`), LF (`\n`) or CRLF (`\r\n`), i.e. `LINES TERMINATED BY`
+    is not customizable.
+* There cannot be line prefixes (`LINES STARTING BY`).
+* The header cannot be simply skipped (`IGNORE n LINES`), it must be valid column names if present.
+* Delimiters and separators can only be a single ASCII character.
+
+## Common configurations
+
+### CSV
+
+The default setting is already tuned for CSV following RFC 4180.
+
+```toml
+[mydumper.csv]
+separator = ','
+delimiter = '"'
+header = true
+not-null = false
+null = '\N'
+backslash-escape = true
+trim-last-separator = false
+```
+
+Example content:
+
+```
+ID,Region,Count
+1,"East",32
+2,"South",\N
+3,"West",10
+4,"North",39
+```
+
+### TSV
+
+```toml
+[mydumper.csv]
+separator = "\t"
+delimiter = ''
+header = true
+not-null = false
+null = 'NULL'
```

Comment:
how can distinguish string `NULL` from NULL value `NULL` ?

Reply:
@amyangfei  Unfortunately you can't, that's why TSV is not a good backup format 🙃. Demo:

```sql
select 'NULL', NULL into outfile '/tmp/1.tsv' fields terminated by "\t" escaped by '';
```

```
NULL	NULL

```